### PR TITLE
[cronet_http] 🏗️ Use dart-define to determine dependency

### LINKED ISSUE #807: cronet_http: depend on org.chromium.net:cronet-api instead
The current implementation depends on `com.google.android.gms:play-services-cronet`, which pulls in the Google Play services. This might be undesirable for apps that do not want to have GMS code in them (and those who mainly target Chinese users and do not have GMS available).

We should depend on Chromium's API package, which will find the relevant dependencies automatically if they are declared. Additional documentations should be provided to the users on how to choose a Cronet provider (similar to https://exoplayer.dev/network-stacks.html#cronet).

### LINKED ISSUE #1047: [cronet_http_embedded] incompatibility with newer versions of cronet_http
`cronet_http` README states that `cronet_http_embedded` is the same library with Cornet bundled. But it hasn't been updated in some time, and these have different interfaces. Since some Chinese manufacturers don't have Google Play Services included, it's vital to bundle cornet.

### LINKED ISSUE #1105: `package:cronet_http_embedded` needs a new name and publisher
@AlexV525 

It is not possible to transfer `package:cronet_http_embedded` to another publisher - this is process limitation for Google-owned packages.

So we need to pick a new name so that it can be published by another publisher, i.e., "flutter.cn". We can update the `package:cronet_http_embedded` package page to point to the new package.

Maybe one of:

- `package:cronet_embedded_http`
- `package:cronet_http_noplay`  # I like this one - it makes it clear why you would use it.
- `package:cronet_http_embed`

But I'm open to other names.

### PR DESCRIPTION
**Pull request by AlexV525**
Instead of creating a new embedded package, we can use `dart-define` to conditionally implement dependencies.
- Fixes #1105
- Fixes #1047
- Permanently fixes #807
- Fixes https://github.com/cfug/dio/issues/2042

521a017a3284d4ca88cb3c63c618785f

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by AlexV525** [comment]
Require these PRs to land first:
- #1091
- #1104

**Comment by AlexV525** [comment]
This also includes #1101.

File: pkgs/cronet_http/README.md [hunk 1]
```diff
@@ -40,22 +38,45 @@ void main() async {
   final Client httpClient;
   if (Platform.isAndroid) {
     final engine = CronetEngine.build(
-        cacheMode: CacheMode.memory,
-        cacheMaxSize: 2 * 1024 * 1024,
-        userAgent: 'Book Agent');
+      cacheMode: CacheMode.memory,
+      cacheMaxSize: 2 * 1024 * 1024,
+       userAgent: 'Book Agent',
```
Comment (brianquinlan) [inline pkgs/cronet_http/README.md:43]: There seems to be some random formatting changes in this file...isn't this formatting actually incorrect?

**Review by brianquinlan — CHANGES_REQUESTED** [review]
This is a really cool approach! I think that you need to increment the pub spec version and add an entry to the CHANGELOG.

One this is live, I think that I can add a note to cronet_http_embedded and deprecate it.

File: pkgs/cronet_http/README.md [hunk 2]
```diff
@@ -40,22 +38,45 @@ void main() async {
...
-      {'q': 'HTTP', 'maxResults': '40', 'printType': 'books'}));
+      {'q': 'HTTP', 'maxResults': '40', 'printType': 'books'},
+    ),
+  );
 }
 ```
 
+### Use embedded Cronet
+
+The embedded [Cronet][] is provided as `org.chromium.net:cronet-embedded`
```
Comment (brianquinlan) [inline pkgs/cronet_http/README.md:62]: Maybe something more explicit like:

By default, `package:cronet_http` uses the [Cronet][] library provided by [Google Play Services][].

If you want your application to work without [Google Play Services][], you can instead depend on the `org.chromium.net:cronet-embedded` package by using `dart-define` to set `cronetHttpNoPlay` is set to `true`.

For example:
...

File: pkgs/cronet_http/README.md [hunk 3]
```diff
@@ -40,22 +38,45 @@ void main() async {
...
+which does not require to use the [Google Play Services][].
+The library uses `dart-define` to switch to the embedded implementation
+when `cronetHttpNoPlay` is set to `true`.
+For example:
+
+```
+flutter run --dart-define=cronetHttpNoPlay=true
```
Comment (brianquinlan) [inline pkgs/cronet_http/README.md:69]: The convention seems to be to use name_with_underscore. So maybe:

cronet_http_no_play=true

**Comment by brianquinlan** [comment]
After patching this in and using it for a bit, I'm even more amazed by this approach!

**Comment by AlexV525** [comment]
@brianquinlan Thanks for the review. I'll update this after #1091 gets merged.

File: pkgs/cronet_http/README.md (hunk 3 above)
Comment (AlexV525) [reply pkgs/cronet_http/README.md:69]: Any guide suggesting this?

File: pkgs/cronet_http/README.md (hunk 1 above)
Comment (AlexV525) [reply pkgs/cronet_http/README.md:43]: Flutter prefers trailing commas, which makes the formatting different.

File: pkgs/cronet_http/README.md (hunk 2 above)
Comment (AlexV525) [reply pkgs/cronet_http/README.md:62]: > By default, `package:cronet_http` uses the [Cronet][] library provided by [Google Play Services][].

The first paragraph duplicates the previous paragraphs. I'll take the second.

File: pkgs/cronet_http/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 1.2.0-wip
```
Comment (brianquinlan) [inline pkgs/cronet_http/CHANGELOG.md:1]: I think that you need to update from master

File: pkgs/cronet_http/README.md (hunk 1 above)
Comment (brianquinlan) [reply pkgs/cronet_http/README.md:43]: This just looks misindented though. Can you change it back?

File: pkgs/cronet_http/README.md (hunk 3 above)
Comment (brianquinlan) [reply pkgs/cronet_http/README.md:70]: No, I just looked at some existing defines e.g. `test_runner.configuration`.

But it is not consistent so feel free to leave it this way.

File: pkgs/cronet_http/README.md (hunk 1 above)
Comment (AlexV525) [reply pkgs/cronet_http/README.md:43]: Sure

File: pkgs/cronet_http/CHANGELOG.md (hunk 4 above)
Comment (AlexV525) [reply pkgs/cronet_http/CHANGELOG.md:1]: Something that I've missed? The master has `1.1.1` as the latest.

Comment (brianquinlan) [reply pkgs/cronet_http/CHANGELOG.md:1]: Sorry, I was expecting some text from an unmerged PR there. My mistake.

**Review by brianquinlan — APPROVED** [review]

**Comment by brianquinlan** [comment]
Thanks!

**Pull request merged**